cilium_eip_no_masquerade_agent/src/main.rs

```diff
@@ -98,36 +81,125 @@ impl k8s_controller::Context for Context {
+                    self.ip_rule_handle.del(rule).execute().await?;
                     break;
                 }
             }
         }
-        Ok(None)
+        self.remove_finalizer(pod, pod_name).await;
+        Ok(())
     }
-}
 
-async fn filter_pod_rules(
-    rules: impl TryStream<Ok = RuleMessage, Error = rtnetlink::Error>,
-    pod_ip: Ipv4Addr,
-) -> Result<Vec<RuleMessage>, rtnetlink::Error> {
-    rules
-        .try_filter(|rule| {
-            future::ready(
-                rule.header.src_len == 32
-                    && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec())),
-            )
-        })
-        .collect::<Vec<Result<RuleMessage, rtnetlink::Error>>>()
-        .await
-        .into_iter()
-        .collect()
+    async fn remove_finalizer(&self, pod: &Pod, pod_name: &str) {
+        // https://docs.rs/kube-runtime/latest/src/kube_runtime/finalizer.rs.html
+        let finalizer_index = pod
+            .finalizers()
+            .iter()
+            .enumerate()
+            .find(|(_, finalizer)| *finalizer == FINALIZER_NAME)
+            .map(|(i, _)| i);
+        if let Some(finalizer_index) = finalizer_index {
+            let pod_api: Api<Pod> =
+                Api::namespaced(self.kube_client.clone(), &pod.namespace().unwrap());
+            let finalizer_path = format!("/metadata/finalizers/{finalizer_index}");
+            pod_api
+                .patch::<Pod>(
+                    pod_name,
+                    &PatchParams::default(),
+                    &Patch::Json(json_patch::Patch(vec![
+                        // All finalizers run concurrently and we use an integer index
+                        // `Test` ensures that we fail instead of deleting someone else's finalizer
+                        PatchOperation::Test(TestOperation {
+                            path: finalizer_path.clone(),
+                            value: FINALIZER_NAME.into(),
+                        }),
+                        PatchOperation::Remove(RemoveOperation {
+                            path: finalizer_path,
+                        }),
+                    ])),
+                )
+                .await
+                .unwrap();
+        }
+    }
+
+    async fn wait_for_chain_to_exist(&self) -> Result<(), Box<dyn std::error::Error>> {
+        info!("Waiting for {CHAIN} chain to exist");
+        while !self.iptables.chain_exists(TABLE, CHAIN)? {
+            trace!("{CHAIN} does not exist, sleeping");
+            sleep(Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    async fn get_eni_indexes_for_existing_ip_rules(
+        &self,
+    ) -> Result<HashSet<u32>, rtnetlink::Error> {
+        let rules = self.ip_rule_handle.get(IpVersion::V4).execute();
+        rules
+            .try_filter(|rule| future::ready(rule.nlas.contains(&Nla::FwMask(0xf))))
+            .map_ok(|rule| {
```

Comment:
nit: I think you should be able to use `try_filter_map` for this, e.g. 

```rust
    .try_filter_map(|rule| async move {
        if !rule.nlas.contains(&Nla::FwMask(0xf)) {
            return None;
        }

        rule.nlas.into_iter().find_map(|nla| {
            match nla {
                Nla::FwMark(i) => Some(i),
                _ => None,
            }
        }).expect("blah")
    })

```

Reply:
I hit the same lifetime issues as above when trying to do an async closure.

Reply:
oh yeah didn't realize rule was borrowed. maybe just `async` and use `iter()` rather than `into_iter()` (and then `*i` rather than `i`)?

Reply:
I got this to compile with:
```rust
        rules
            .try_filter_map(|rule| async {
                if !rule.nlas.contains(&Nla::FwMask(FW_MASK)) {
                    return Ok(None);
                }
                Ok(rule.nlas.into_iter().find_map(|nla| match nla {
                    Nla::FwMark(i) => Some(i),
                    _ => None,
                }))
            })
            .try_collect()
            .await
```
but this subtly changes the behavior. Previously, we would panic if we hit a rule that had our mask, but no mark. I actually think the panic was more appropriate here. I guess I could use a custom error type, but there isn't really anything to do about this other than fail.